### PR TITLE
Upstash integration backend

### PR DESCRIPTION
**Pull request by ferozemohideen**
following instructions from https://upstash.com/docs/common/help/integration#technical-design-gui-apps

steps to creating integration:
1. direct user to porter project middleware, inject state and project into session then redirect to auth url
2. user logs in, then hits the callback url
3. callback handler gets an access token, then uses that to get a developer token. Developer token is stored in db and never expires

### CODE REVIEW

File: api/server/handlers/oauth_callback/upstash.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+		telemetry.AttributeKV{Key: "project-id", Value: projID},
+	)
+
+	if projID == 0 {
+		err = telemetry.Error(ctx, span, nil, "project id not found in session")
+		p.HandleAPIError(w, r, apierrors.NewErrPassThroughToClient(err, http.StatusInternalServerError))
+		return
+	}
+
+	token, err := p.Config().UpstashConf.Exchange(ctx, r.URL.Query().Get("code"))
```
Comment (stefanmcshane) [inline api/server/handlers/oauth_callback/upstash.go:78]: Will this complain if `code` is empty?

File: api/server/handlers/oauth_callback/upstash.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	}
+
+	oauthInt, err = p.Repo().UpstashIntegration().Insert(ctx, oauthInt)
+	if err != nil {
+		err = telemetry.Error(ctx, span, err, "error creating oauth integration")
+		p.HandleAPIError(w, r, apierrors.NewErrPassThroughToClient(err, http.StatusInternalServerError))
+		return
+	}
+
+	if redirectStr, ok := session.Values["redirect_uri"].(string); ok && redirectStr != "" {
```
Comment (stefanmcshane) [inline api/server/handlers/oauth_callback/upstash.go:117]: Not a blocker - else could be dropped

```go
redirect := "/dashboard"
if redirectStr, ok := session.Values["redirect_uri"].(string); ok && redirectStr != "" {
    redirectURI, err := url.Parse(redirectStr)
    if err == nil {
      redirect := fmt.Sprintf("%s?%s", redirectURI.Path, redirectURI.RawQuery)}
    }
}
http.Redirect(w, r, redirect, http.StatusFound)
```

File: api/server/handlers/oauth_callback/upstash.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, UpstashApiKeyEndpoint, bytes.NewBuffer(jsonData))
+	if err != nil {
+		return "", telemetry.Error(ctx, span, err, "error creating request")
+	}
+
+	// Set the Authorization header
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", accessToken))
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := client.Do(req)
```
Comment (stefanmcshane) [inline api/server/handlers/oauth_callback/upstash.go:163]: If you arent doing anything with the client, you can use `http.DefaultClient.Do()`

File: api/server/handlers/oauth_callback/upstash.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	if err != nil {
+		return "", telemetry.Error(ctx, span, err, "error sending request")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		telemetry.WithAttributes(span, telemetry.AttributeKV{Key: "status-code", Value: resp.StatusCode})
```
Comment (stefanmcshane) [inline api/server/handlers/oauth_callback/upstash.go:170]: maybe add this regardless.

can `response-body` contain anything sensitive?

File: api/server/shared/config/config.go [hunk 5]
```diff
@@ -76,6 +76,9 @@ type Config struct {
 	// SlackConf is the configuration for a Slack OAuth client
 	SlackConf *oauth2.Config
 
+	// UpstashConf is the configuration for an Upstash OAuth client
+	UpstashConf *oauth2.Config
```
Comment (stefanmcshane) [inline api/server/shared/config/config.go:80]: Can we remove the pointer?

File: api/server/shared/config/env/envconfs.go [hunk 6]
```diff
@@ -82,6 +82,8 @@ type ServerConf struct {
 	SlackClientID     string `env:"SLACK_CLIENT_ID"`
 	SlackClientSecret string `env:"SLACK_CLIENT_SECRET"`
 
+	UpstashClientID string `env:"UPSTASH_CLIENT_ID"`
```
Comment (stefanmcshane) [inline api/server/shared/config/env/envconfs.go:85]: Its good practice to specifically enable these things .i.e `UPSTASH_ENABLED`

That way, you can fail on startup if upstash should be enabled, but has no clientID etc

File: zarf/helm/.serverenv [hunk 7]
```diff
@@ -81,3 +81,6 @@ PORTER_CLOUD_PLAN_ID=
 
 # PORTER_STANDARD_PLAN_ID is the id of the standard plan in Metronome. Only used if METRONOME_API_KEY is set
 PORTER_STANDARD_PLAN_ID=
+
+# UPSTASH_CLIENT_ID is used to integrate with Upstash
```
Comment (stefanmcshane) [inline zarf/helm/.serverenv:85]: How can we get one of these?

**Review by stefanmcshane — APPROVED** [review]

File: zarf/helm/.serverenv (hunk 7 above)
Comment (ferozemohideen) [reply zarf/helm/.serverenv:85]: upstash team sent me one in plain text over slack lol

**Pull request merged**